### PR TITLE
Fix assembly conflict issue

### PR DESCRIPTION
**Pull request by Shwetap05**
This PR brings lots of breaking changes

- Added logic to handle assembly load conflicts that would work for both .Net Framework and .Net Core run times.
- All of the type load logic is moved up so all the required type information is fetched before we actually proceed with generation.
- Updated contracts for filters to include list of generation errors, that would be populated by each filter as it processes various xElements.
- Added newtonsoft to handle serialization.
- Other cascading changes.
- Updated major version of the package.

### CODE REVIEW

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/InternalOpenApiGenerator.cs [hunk 1]
```diff
@@ -305,10 +293,40 @@ internal class InternalOpenApiGenerator
                     });
                 }
 
-                var typeFetcher = new TypeFetcher(contractAssemblyPaths);
+#if !NETFRAMEWORK
+                var assemblyLoader = new AssemblyLoader.AssemblyLoader();
+                assemblyLoader.RegisterAssemblyPaths(contractAssemblyPaths);
+                var stringSchemaTypeInfo = new AssemblyLoader.AssemblyLoader().BuildSchemaTypeInfo(
```
Comment (Shwetap05) [inline src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/InternalOpenApiGenerator.cs]: >SchemaTypeInfo [](start = 84, length = 14)

will update naming around SchemaTypeInfo to GenerationContext

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/AssemblyLoader/CustomAssemblyLoadContext.cs [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+            return null;
+        }
+
+        private Assembly TryLoadByAssemblyName(string assemblyName)
+        {
+            if (!_assembliesLoadedByAssemblyName.Contains(assemblyName))
+            {
+                try
+                {
+                    _assembliesLoadedByAssemblyName.Add(assemblyName);
```
Comment (rwrife) [inline src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/AssemblyLoader/CustomAssemblyLoadContext.cs]: nit. whitespace around the ( )

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/Extensions/IListExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration.Extensions
+{
+    /// <summary>
+    /// 
+    /// </summary>
+    internal static class IListExtensions
+    {
+        /// <summary>
+        ///
```
Comment (rwrife) [inline src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/Extensions/IListExtensions.cs]: nit. empty summaries

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/OperationFilters/RemarksToDescriptionFilter.cs [hunk 4]
```diff
@@ -22,26 +25,46 @@ public class RemarksToDescriptionFilter : IOperationFilter
         /// <param name="operation">The operation to be updated.</param>
         /// <param name="element">The xml element representing an operation in the annotation xml.</param>
         /// <param name="settings">The operation filter settings.</param>
+        /// <returns>The list of generation errors, if any produced when processing the filter."></returns>
```
Comment (rwrife) [inline src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/OperationFilters/RemarksToDescriptionFilter.cs]: remove "> from the end

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/OperationFilters/ResponseToResponseFilter.cs [hunk 5]
```diff
@@ -24,154 +29,184 @@ public class ResponseToResponseFilter : IOperationFilter
         /// <param name="operation">The operation to be updated.</param>
         /// <param name="element">The xml element representing an operation in the annotation xml.</param>
         /// <param name="settings">The operation filter settings.</param>
+        /// <returns>The list of generation errors, if any produced when processing the filter."></returns>
```
Comment (rwrife) [inline src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/OperationFilters/ResponseToResponseFilter.cs]: remove ">

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/OperationFilters/SummaryToSummaryFilter.cs [hunk 6]
```diff
@@ -22,25 +25,45 @@ public class SummaryToSummaryFilter : IOperationFilter
         /// <param name="operation">The operation to be updated.</param>
         /// <param name="element">The xml element representing an operation in the annotation xml.</param>
         /// <param name="settings">The operation filter settings.</param>
+        /// <returns>The list of generation errors, if any produced when processing the filter."></returns>
```
Comment (rwrife) [inline src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/OperationFilters/SummaryToSummaryFilter.cs]: remove ">

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/PostProcessingDocumentFilters/RemoveFailedGenerationOperationFilter.cs [hunk 7]
```diff
@@ -20,41 +22,58 @@ public class RemoveFailedGenerationOperationFilter : IPostProcessingDocumentFilt
         /// </summary>
         /// <param name="openApiDocument">The OpenAPI document to process.</param>
         /// <param name="settings">The filter settings.</param>
-        public void Apply(
+        /// <returns>The list of generation errors, if any produced when processing the filter."></returns>
```
Comment (rwrife) [inline src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/PostProcessingDocumentFilters/RemoveFailedGenerationOperationFilter.cs]: ">

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/PreprocessingOperationFilters/BranchOptionalPathParametersFilter.cs [hunk 8]
```diff
@@ -26,46 +27,66 @@ public class BranchOptionalPathParametersFilter : IPreProcessingOperationFilter
         /// <param name="paths">The paths to be updated.</param>
         /// <param name="element">The xml element representing an operation in the annotation xml.</param>
         /// <param name="settings">The operation filter settings.</param>
-        public void Apply(OpenApiPaths paths, XElement element, PreProcessingOperationFilterSettings settings)
+        /// <returns>The list of generation errors, if any produced when processing the filter."></returns>
```
Comment (rwrife) [inline src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/PreprocessingOperationFilters/BranchOptionalPathParametersFilter.cs]: ">

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/PreprocessingOperationFilters/ConvertAlternativeParamTagsFilter.cs [hunk 9]
```diff
@@ -21,123 +24,144 @@ public class ConvertAlternativeParamTagsFilter : IPreProcessingOperationFilter
         /// <param name="paths">The paths to be updated.</param>
         /// <param name="element">The xml element representing an operation in the annotation xml.</param>
         /// <param name="settings">The operation filter settings.</param>
-        public void Apply(OpenApiPaths paths, XElement element, PreProcessingOperationFilterSettings settings)
+        /// <returns>The list of generation errors, if any produced when processing the filter."></returns>
```
Comment (rwrife) [inline src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/PreprocessingOperationFilters/ConvertAlternativeParamTagsFilter.cs]: ">

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/PreprocessingOperationFilters/IPreProcessingOperationFilter.cs [hunk 10]
```diff
@@ -26,6 +28,7 @@ public interface IPreProcessingOperationFilter : IFilter
         /// <see cref="OpenApiGeneratorConfig.AnnotationXmlDocuments"/>.
         /// </param>
         /// <param name="settings"><see cref="PreProcessingOperationFilterSettings"/></param>
-        void Apply(OpenApiPaths paths, XElement element, PreProcessingOperationFilterSettings settings);
+        /// <returns>The list of generation errors, if any produced when processing the filter."></returns>
```
Comment (rwrife) [inline src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/PreprocessingOperationFilters/IPreProcessingOperationFilter.cs]: ">

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/PreprocessingOperationFilters/PopulateInAttributeFilter.cs [hunk 11]
```diff
@@ -25,48 +27,68 @@ public class PopulateInAttributeFilter : IPreProcessingOperationFilter
         /// <param name="paths">The paths to be updated.</param>
         /// <param name="element">The xml element representing an operation in the annotation xml.</param>
         /// <param name="settings">The operation filter settings.</param>
-        public void Apply(OpenApiPaths paths, XElement element, PreProcessingOperationFilterSettings settings)
+        /// <returns>The list of generation errors, if any produced when processing the filter."></returns>
```
Comment (rwrife) [inline src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/PreprocessingOperationFilters/PopulateInAttributeFilter.cs]: ">

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/AssemblyLoader/AssemblyLoadUtility.cs [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+                {
+                    var info = FileVersionInfo.GetVersionInfo(assemblyPath);
+                    if (info.FileVersion.StartsWith(assemblyVersion))
+                    {
+                        return Assembly.LoadFrom(assemblyPath);
+                    }
+                }
+                catch (Exception)
+                {
+                    // Do nothing.
```
Comment (rwrife) [inline src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/AssemblyLoader/AssemblyLoadUtility.cs]: maybe I'm wrong, but is this not a big exception that should halt the entire operation?

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/AssemblyLoader/AssemblyLoader.cs [hunk 13]
```diff
@@ -0,0 +1,412 @@
+﻿// ------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (Shwetap05) [reply src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/AssemblyLoader/AssemblyLoader.cs]: Yes i meant to do that, forgot.

---
In reply to: [284032234](https://github.com/microsoft/OpenAPI.NET.CSharpAnnotations/pull/208#discussion_r284032234) [](ancestors = 284032234)

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/DocumentConfigFilters/DocumentVariantAttributesFilter.cs [hunk 14]
```diff
@@ -30,70 +31,86 @@ public class DocumentVariantAttributesFilter : IDocumentConfigFilter
...
-            var variantElements = element.Elements(KnownXmlStrings.Variant);
-            
-            foreach (var variantElement in variantElements)
+            var generationErrors = new List<GenerationError>();
+
+            try
             {
-                // First, populate the document variant info attributes using the document variant option.
-                var variantOptions = variantElement.Element(KnownXmlStrings.Options)?.Elements(KnownXmlStrings.Option);
+                var variantElements = element.Elements(KnownXmlStrings.Variant);
```
Comment (rwrife) [inline src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/DocumentConfigFilters/DocumentVariantAttributesFilter.cs]: do you need a null check on the element?

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/Models/InternalGenerationContext.cs [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+            {
+                CrefToFieldValueMap = this.CrefToFieldValueMap
+            };
+
+            foreach (var key in this.CrefToSchemaMap.Keys)
+            {
+                InternalSchemaGenerationInfo schemaInfo = this.CrefToSchemaMap[key];
+                var internalSchemaInfo = new SchemaGenerationInfo
+                {
+                    Schema = schemaInfo.Schema == null
```
Comment (rwrife) [inline src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/Models/InternalGenerationContext.cs]: minor readability issue with the ternary, consider:
Schema = (schemaInfo.Schema == null)
   ? blah
   : other blah;

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/AssemblyLoader/CustomAssemblyLoadContext.cs (hunk 2 above)
Comment (Shwetap05) [reply src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/AssemblyLoader/CustomAssemblyLoadContext.cs]: Yes i had to disable resharper due to conflicting guidelines internal and externally, will update this.

---
In reply to: [284022572](https://github.com/microsoft/OpenAPI.NET.CSharpAnnotations/pull/208#discussion_r284022572) [](ancestors = 284022572)

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/AssemblyLoader/CustomAssemblyLoadContext.cs [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+            if (version != null)
+            {
+                var assemblyByVersion = TryLoadByVersion(AssemblyPaths, assemblyName, version.Major + "." + version.Minor + "." + version.Build + ".");
+                if (assemblyByVersion != null)
+                {
+                    Assemblies[args.Name] = assemblyByVersion;
+                    return assemblyByVersion;
+                }
+
+                assemblyByVersion = TryLoadByVersion(AssemblyPaths, assemblyName, version.Major + "." + version.Minor + ".");
```
[the remark this answers is not in the document]
Comment (Shwetap05) [reply src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/AssemblyLoader/CustomAssemblyLoadContext.cs]: how do you recommend fixing this? b/c first i try to fetch with full version and then with minor and then just major.

---
In reply to: [284004194](https://github.com/microsoft/OpenAPI.NET.CSharpAnnotations/pull/208#discussion_r284004194) [](ancestors = 284004194)

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/PreprocessingOperationFilters/PopulateInAttributeFilter.cs (hunk 11 above)
Comment (Shwetap05) [reply src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/PreprocessingOperationFilters/PopulateInAttributeFilter.cs]: Sure will fix them all.

---
In reply to: [284025876](https://github.com/microsoft/OpenAPI.NET.CSharpAnnotations/pull/208#discussion_r284025876) [](ancestors = 284025876)

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/AssemblyLoader/AssemblyLoader.cs [hunk 17]
```diff
@@ -0,0 +1,412 @@
...
+                    }
+                }
+
+                var assembly = AssemblyLoadUtility.TryLoadByName(assemblyPaths, assemblyName);
+                if (assembly != null)
+                {
+                    return assembly;
+                }
+
+                assembly = AssemblyLoadUtility.TryLoadByName(assemblyPaths, assemblyName);
```
[the remark this answers is not in the document]
Comment (Shwetap05) [reply src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/AssemblyLoader/AssemblyLoader.cs]: this look s like copy paste error

---
In reply to: [282709466](https://github.com/microsoft/OpenAPI.NET.CSharpAnnotations/pull/208#discussion_r282709466) [](ancestors = 282709466)

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/AssemblyLoader/AssemblyLoadUtility.cs [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+        /// <param name="assemblyPaths">The assembly paths.</param>
+        /// <param name="assemblyName">The assembly name to load.</param>
+        /// <param name="assemblyVersion">The assembly version to load.</param>
+        /// <returns>The assembly if able to load, otherwise null.</returns>
+        internal static Assembly TryLoadByVersion(
+            IList<string> assemblyPaths,
+            string assemblyName,
+            string assemblyVersion)
+        {
+            var assemblyPath = assemblyPaths.FirstOrDefault(path => path.EndsWith(assemblyName + ".dll"));
```
[the remark this answers is not in the document]
Comment (Shwetap05) [reply src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/AssemblyLoader/AssemblyLoadUtility.cs]: b/c we are checking against assembly path, so path would be like //test/test.dll and we watch to match test.dll with the path, so ends with.

---
In reply to: [282706378](https://github.com/microsoft/OpenAPI.NET.CSharpAnnotations/pull/208#discussion_r282706378) [](ancestors = 282706378)

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/DocumentConfigFilters/DocumentVariantAttributesFilter.cs (hunk 14 above)
Comment (Shwetap05) [reply src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/DocumentConfigFilters/DocumentVariantAttributesFilter.cs]: sure will do on all parameters

---
In reply to: [284036413](https://github.com/microsoft/OpenAPI.NET.CSharpAnnotations/pull/208#discussion_r284036413) [](ancestors = 284036413)

File: src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/AssemblyLoader/CustomAssemblyLoadContext.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Shwetap05) [reply src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/AssemblyLoader/CustomAssemblyLoadContext.cs]: but two assemblies with different casing could not be under same path right, so will never run into that issue right? or am i missing something here?

---
In reply to: [284005872](https://github.com/microsoft/OpenAPI.NET.CSharpAnnotations/pull/208#discussion_r284005872) [](ancestors = 284005872)

**Pull request merged**